REST endpoints for file creation

adamdecaf opened the pull request:
Creates a REST interface for building ACH files. This is also going to be our first deployed service. 

Issue: https://github.com/moov-io/ach/issues/254

wadearnold committed:
File CRUD functions and tests

wadearnold committed:
File CRUD functions and tests

wadearnold committed:
Basic CRUD File endpoints

wadearnold committed:
Basic CRUD File endpoints

wadearnold committed:
Merge remote-tracking branch 'moov-io/master' into REST-Server

wadearnold committed:
Merge remote-tracking branch 'moov-io/master' into REST-Server

wadearnold committed:
Merge remote-tracking branch 'moov-io/master' into REST-Server

wadearnold committed:
Merge branch 'master' into REST-Server

wadearnold committed:
Merge branch 'REST-Server' of https://github.com/wadearnold/ach into REST-Server

wadearnold committed:
Merge branch 'master' into REST-Server

wadearnold committed:
Ignore /server/ in travis build

wadearnold committed:
Remove NOC and Returns if they don't exist

wadearnold committed:
OpenAPI specification

Create initital endpoint mapping in OpenAPI for working with a File

wadearnold committed:
Merge branch 'master' into REST-Server

wadearnold committed:
Merge branch 'master' into REST-Server

wadearnold committed:
Batch enpoint start

wadearnold committed:
template batch

wadearnold committed:
import reference to moov-io

wadearnold committed:
Merge branch 'master' into REST-Server

wadearnold committed:
Mock service enpoints

wadearnold committed:
GetBatch completed

wadearnold committed:
add store.findBatch & service.GetBatch

wadearnold committed:
Support GetBatch in transport

wadearnold committed:
GetBatches & DeleteBatches

wadearnold committed:
Merge branch 'master' into REST-Server

wadearnold committed:
Merge conflict resolution

wadearnold committed:
Merge branch 'master' into REST-Server

wadearnold committed:
Remove Batch from File on DeleteBatch

wadearnold committed:
Merge branch 'master' into REST-Server

adamdecaf commented on cmd/server/main.go:
```diff
@@ -0,0 +1,79 @@
+	"syscall"
+
+	"github.com/wadearnold/ach/server"
+	"github.com/wadearnold/kit/log"
```
fix import path

adamdecaf commented on server/repositoryInMemory.go:
```diff
@@ -0,0 +1,123 @@
+	DeleteBatch(fileID string, batchID string) error
+}
+type repositoryInMemory struct {
+	mtx   sync.RWMutex
```
@wadearnold a bit of a style thing, but I sometimes embed right into the struct (but you expose Lock/Unlock publically here. Since `repositoryInMemory` isn't exported though we don't have to worry as much about letting loose implementation details.

```Go
type Foo struct {
  sync.Mutex
}

// func (f *Foo) Action() error { // calls f.Lock() / f.Unlock()

// vs

type Bar struct {
  
}

wadearnold replied:
Thanks for the feedback. I never would have thought about how someone could just use RepositoryInMemory.mtx from outside the package. Habits!

adamdecaf replied:
They won't be able to use `mtx` as it's lowercase (not exported) nor are they able to use `repositoryInMemory`.

adamdecaf commented on server/service.go:
```diff
@@ -0,0 +1,144 @@
+	// DeleteFile takes a file resource ID and deletes it from the store
+	DeleteFile(id string) error
+	// UpdateFile updates the changes properties of a matching File ID
+	// UpdateFile(f ach.File) (string, error)
```
TODO(adam)

adamdecaf committed:
Merge remote-tracking branch 'origin/master' into REST-Server

adamdecaf committed:
server: fix import paths

adamdecaf committed:
Merge branch 'modules' into REST-Server

adamdecaf committed:
fixup uuid dep, add vendor/

adamdecaf committed:
build: cleanup go calls for go 1.10+

'go vet' is ran by 'go test' in 1.10+ (and skips vendor/ dirs)

adamdecaf committed:
cmd/server: misc cleanup to better Go

adamdecaf committed:
server: create IDs in a more uniform way

Taking part of a UUID isn't the best for unique IDs as often slugs in
a UUID won't vary across multiple calls.

Instead, let's read some random bytes, generate a checksum and take part of that.

adamdecaf committed:
server: ensure ID's are 16 characters

adamdecaf committed:
server: rework to pass around pointers rather than copying structs

adamdecaf committed:
server: don't panic, ask user to report issue

adamdecaf committed:
server: clarify remaining endpoints

adamdecaf committed:
server: report timings in milliseconds

adamdecaf committed:
go fmt

adamdecaf committed:
Merge pull request #1 from adamdecaf/enhance-log-timing

server: report timings in milliseconds

adamdecaf committed:
go fmt

adamdecaf committed:
doc: link to swagger docs

adamdecaf committed:
go fmt

adamdecaf committed:
server: add GET /files/:id/validate

Issue: https://github.com/moov-io/ach/issues/254

adamdecaf committed:
server: drop incorrect "format: uuid" for params, fix spelling

The ID's were never really UUID's, unless the user had provided them.

adamdecaf committed:
server: add POST /files/ that reads body as plaintext ACH file

Issue: https://github.com/moov-io/ach/issues/254

adamdecaf committed:
server: rename POST /files to /files/create

Issue: https://github.com/moov-io/ach/issues/254

adamdecaf committed:
server: add PATCH /files/:id/build

Issue: https://github.com/moov-io/ach/issues/254

adamdecaf committed:
server: use ValidateFile in tests, not file level Validation

adamdecaf committed:
cmd/server: fix crude curl docs

adamdecaf committed:
server: get a file's plaintext via GET /files/:id/contents

adamdecaf committed:
server: consolidate on error json response format

adamdecaf committed:
server: add quick sanity spec for Repository

adamdecaf committed:
Merge remote-tracking branch 'origin' into REST-Server

adamdecaf committed:
cmd/server: add basic hardening for net/http on the internet

Source: https://www.youtube.com/watch?v=wxkEQxvxs3w
